packages/engine/paima-funnel/src/funnels/avail/utils.ts

```diff
@@ -0,0 +1,171 @@
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const babeDigest = rawBabeDigest.toPrimitive() as unknown as any;
+
+  // the object is an enumeration, but all the variants have a slotNumber field
+  const slot = babeDigest[Object.getOwnPropertyNames(babeDigest)[0]].slotNumber;
+  return slot;
+}
+
+export async function getLatestBlockNumber(api: ApiPromise): Promise<number> {
+  let highHash = await api.rpc.chain.getFinalizedHead();
+  let high = (await api.rpc.chain.getHeader(highHash)).number.toNumber();
+  return high;
+}
+
+export async function getTimestampForBlockAt(api: ApiPromise, mid: number): Promise<number> {
+  const hash = await api.rpc.chain.getBlockHash(mid);
+  // FIXME: why is the conversion needed?
+  const header = (await api.rpc.chain.getHeader(hash)) as unknown as Header;
+
+  const slot = getSlotFromHeader(header, api);
+  return slotToTimestamp(slot, api);
+}
+
+export function slotToTimestamp(slot: number, api: ApiPromise): number {
+  // this is how it's computed by the pallet
+  // https://paritytech.github.io/polkadot-sdk/master/src/pallet_babe/lib.rs.html#533
+  const slotDuration = (Number.parseInt(api.consts.timestamp.minimumPeriod.toString()) * 2) / 1000;
+
+  // slots start at unix epoch:
+  // https://paritytech.github.io/polkadot-sdk/master/src/pallet_babe/lib.rs.html#902
+  return slot * slotDuration;
+}
+
+// inverse to `slotToTimestamp`
+export function timestampToSlot(timestamp: number, api: ApiPromise): number {
+  const slotDuration = (Number.parseInt(api.consts.timestamp.minimumPeriod.toString()) * 2) / 1000;
+
+  // slots start at the unix epoch regardless of the genesis timestamp
+  return timestamp / slotDuration;
+}
+
+type HeaderData = { number: number; hash: string; slot: number };
+
+export async function getMultipleHeaderData(
+  api: ApiPromise,
+  blockNumbers: number[]
+): Promise<HeaderData[]> {
+  const results = [] as HeaderData[];
+
+  for (const bn of blockNumbers) {
+    // NOTE: the light client allows getting header directly from block number,
+    // but it doesn't provide the babe data for the slot
+    const hash = await api.rpc.chain.getBlockHash(bn);
+    const header = (await api.rpc.chain.getHeader(hash)) as unknown as Header;
+
+    const slot = getSlotFromHeader(header, api);
+
+    results.push({
+      number: header.number.toNumber(),
+      hash: header.hash.toString(),
+      slot: slot,
+    });
+  }
+
+  return results;
+}
+
+export async function getDAData(
+  api: ApiPromise,
+  lc: string,
+  from: number,
+  to: number
+): Promise<{ blockNumber: number; submittedData: SubmittedData[] }[]> {
+  const data = [] as { blockNumber: number; submittedData: SubmittedData[] }[];
+
+  for (let curr = from; curr <= to; curr++) {
+    const responseRaw = await fetch(`${lc}/v2/blocks/${curr}/data?fields=data,extrinsic`);
+
+    // TODO: handle better the status code ( not documented in the api though ).
```

Comment:
It would be good to create an issue on the Avail repo/docs to mention documenting this if you think it would help, then linking to that issue from this code

Reply:
Makes sense. There are basically two issues here:

1. I didn't see any docs for the status code. Not much of an issue because I can read the code, but it makes me a bit uncertain about how it could change.
2. If a block doesn't have _any_ submitted data (from you or from anyone) it returns a 400 (bad request) instead of a 404. This means we need to treat bad request the same as 404 (which is returned if you ask for a block in the future I think, but we should never do that).

Reply:
Okay, so do you want to create an issue and link to it from our code?

Reply:
done